### ISSUE TITLE
Top-level warnings show wrong code

### ISSUE THREAD

**Issue by jdemeyer** [ipython/ipython#11188]
Raising a warning from the top level in IPython:
```
Python 3.6.1 (default, Jun 14 2018, 14:37:55) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: if True:
   ...:     from warnings import warn
   ...:     warn("BOO!")
   ...:     
/usr/local/src/sage-config/local/bin/ipython3:3: UserWarning: BOO!
  # -*- coding: utf-8 -*-
```

A warning is raised from line 3 of the code entered in IPython but line 3 of the `.../bin/ipython3` script is shown. Python itself does this better:
```
>>> if True:
...     from warnings import warn
...     warn("BOO!")
... 
__main__:3: UserWarning: BOO!
```

You may argue that raising warnings from the top level is just silly, but the same might happen when calling a function which raises a warning with a higher `stacklevel` or when the warning is raised in Cython (the latter could be considered a Cython bug).

See also https://github.com/iminuit/iminuit/issues/232

**Comment by njsmith** [comment]
Pretty sure this is a python bug: https://www.bugs.python.org/issue33375

I'm not sure, but it looks like the fix missed 3.7, and has been merged for 3.8.

**Comment by jdemeyer** [comment]
I agree. This is really a duplicate of #11080.
